### ISSUE TITLE
Quick Update for offline stored caches

### ISSUE THREAD

**Issue by wolverine007** [cgeo/cgeo#9595]
Hello,

I am storing all nearby unfound caches as an offline copy, so the whole data is available while clicking on a cache on the live map.

From time to time I am updating this list, to receive the current status and latest logs. With several hundert caches this might take a while. 

So I kindly ask to implement the following features:

1) Quick Update function which is only grabbing the cache status, so I am afterwards able to delete archived Caches from the lists, before updating the whole list 

2) Update offline Cache data in the background via a scheduled task or just by putting the task in the background, right now the task has to stay in front of anything else, elsewise it will be stopped in total.

3) Update data of offline Caches where the stored data is older than a specified date

4) Update data of offline cache in parallel, so cgeo is grabbing the fresh data for a given range of caches simultaneously.

Thanks and stay healthy, 
Wolv

**Comment by geocachermgo** [comment]
I guess the time effort is not to handle and store it but to get the website. So if you have it you can extract all information from it. To just get the state may not save time compared to getting the first logs.

**Comment by wolverine007** [comment]
Might be the issue here... therefore a tasked update during night hours or a task running in the background might be helpful. 

But it seems that the live map is grabbing the status somehow, cause some caches are displayed deactivated, while the latest logs are not stating the deactivation message which has taken place meanwhile.

**Comment by wolverine007** [comment]
Any idea or progress here?

cgeo is definitely showing the correct (current) cache status inside the pop up when you are touching a saved cache on the live map.

Pushing details is opening the full cache description but here the stored status is shown.

So it is still seems that cgeo is capable to do a quick refresh of the cache status.

**Comment by SammysHP** [comment]
We already have an issue (very old) for background tasks.

Partial updates are a bad idea because they cause inconsistent states. I wonder why the popup uses different data in your example.

**Comment by fm-sys** [comment]
The live map API returns also the found state. Therefore we show this data in the cache pop-up, but we do not store these (new) founds states automatically in the DB...

**Comment by SammysHP** [comment]
Reference for background tasks: #2068

**Comment by wolverine007** [comment]
Not sure why the issue has now been linked to a topic from 8y ago which is as well closed 🔐...

Whatever, downloading offline maps in the background is possible, is this technique somehow adoptable for refreshing the offline stored caches?

Coming back to the fact that the popup is showing the correct status, could this api status been grabbed by a refresh mechanism, so at least archived caches could be easily found and sorted out?

But you are right a full refresh of the offline stored caches is much better, but sitting in front of the device and watching a progress bar being processed for minutes or hours is a pain in the back

**Comment by Lineflyer** [comment]
> Coming back to the fact that the popup is showing the correct status, could this api status been grabbed by a refresh mechanism, so at least archived caches could be easily found and sorted out?

We discussed this some while ago and carefully selected which status to take over. Its a bit complicated to make the user understand whats going on in such cases (e.g. you would have the archive status but no corresponding archive log). Such missing data might even lead to inconsistency at other places in c:geo (e.g. when a certain filter or method relies on a combination of infos or if some states are not matching).  
Especially when it comes to merging data from cgeo database with information coming from the map we experienced tons of issues over the years as the live map data turned out not to be up-to-date and correct all the time. 

Summed up it would be IMHO more wise to either enable refreshing in the background and/or to implement a lite refresh (e.g. not reloading any pictures or other elements which takes most time in the refresh process).

It might even be, that such an "light refresh" approach does only makes sense if c:geo has a more reliable interface (e.g. the API to gc.com)....if that might come or not...who knows.

**Comment by wolverine007** [comment]
Hi all,

any further information about a background refresh function?

Happy Easter to the whole team
wolv_007

**Comment by moving-bits** [comment]
Hi @wolverine007,

please keep in mind that we are a (small) team of volunteers doing all work on c:geo in our spare time, with very limited time only. Currently we are working on two pretty heavy tasks, finishing migration to "Android storage access framework" (already on `release` branch, in preparation of next release), and integrating a routing engine (currently on `master`), which keep us pretty busy currently. So it might take some more weeks or months to get your feature request implemented. (Or it might even be that it is never implemented at all, if none of the devs jump to it... This is worst case, for sure, just wanted to explain that it is non-deterministic.)

**Comment by wolverine007** [comment]
Hi @moving-bits ,

thanks for your reply, so as long as an issue is open it might be handled by someone of the team?
Fine for me, I will handle so far the updates over night while my phone is charging and not needed.

Thanks for all the effort you and the others spend to improve the app from release to release, I really appreciate it 👍

Cheers wolv_007

**Comment by moving-bits** [comment]
> so as long as an issue is open it might be handled by someone of the team?

Yes, or at least no one has declined it actively. That's why we keep those issues open, even long running ones. (We even have some three-digit items left...)

If we think an opened issue should be rejected for any reason, we will communicate it in the issue. 
Same if we think something should be discussed first - we will post it in the issue then.

**Comment by fm-sys** [comment]
Isn't it the same as requested in #330?

**Comment by moving-bits** [comment]
> Isn't it the same as requested in #330?

Partly. We had a separate issue somewhere asking to sync current status from live map, without downloading the whole cache details (or just the logs) again - don't have the issue number at hand (maybe #10914, #10924 - but I think there was an older one). That would suffice for item number 1, whereas #330 goes into direction of item number 2 (with certain variations in the other items).

**Comment by wolverine007** [comment]
The app is meanwhile offering to download the routing and map updates in the background, could this process being used to trigger the update of a full local list of caches for instance?


BTW: Installed Android Studio on my laptop, everything working fine, except the case that the app is loading (logo displayed) and after a few sec. directly closing on my virtual android device. Any clue what might be wrong?

**Comment by fm-sys** [comment]
> BTW: Installed Android Studio on my laptop, everything working fine, except the case that the app is loading (logo displayed) and after a few sec. directly closing on my virtual android device. Any clue what might be wrong?

Is there any error in the log output ("run" tab at the bottom of Android studio)? Hard to say otherwise... 

> The app is meanwhile offering to download the routing and map updates in the background, could this process being used to trigger the update of a full local list of caches for instance?

The download process itself is a bit different, as the refresh of a list is not like downloading a single big file.

But `AbstractForegroundIntentService` can probably be easily reused. 

https://github.com/cgeo/cgeo/blob/1d455c29df34143a19302ee2a9bfafc0f1ffb07f/main/src/cgeo/geocaching/service/AbstractForegroundIntentService.java

**Comment by wolverine007** [comment]
@fm-sys below the output from the run tab. created the virtual Pixel right before the test to exclude left overs of my other tries as a possible problem.

09/21 20:03:26: Launching 'main' on Pixel XL API 29.
App restart successful without requiring a re-install.
$ adb shell am start -n "cgeo.geocaching/cgeo.geocaching.SplashActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Connected to process 7612 on device 'Pixel_XL_API_29 [emulator-5556]'.
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
I/cgeo.geocachin: The ClassLoaderContext is a special shared library.
W/cgeo.geocachin: Accessing hidden method Landroid/view/WindowManagerGlobal;->getInstance()Landroid/view/WindowManagerGlobal; (greylist, reflection, allowed)
W/cgeo.geocachin: Accessing hidden field Landroid/view/WindowManagerGlobal;->mViews:Ljava/util/ArrayList; (greylist, reflection, allowed)
W/cgeo.geocachin: Accessing hidden field Landroid/app/ActivityThread;->mH:Landroid/app/ActivityThread$H; (greylist, reflection, allowed)
    Accessing hidden method Landroid/app/ActivityThread;->currentActivityThread()Landroid/app/ActivityThread; (greylist, reflection, allowed)
    Accessing hidden field Landroid/os/Handler;->mCallback:Landroid/os/Handler$Callback; (greylist, reflection, allowed)
W/cgeo.geocachin: Accessing hidden field Landroid/util/Singleton;->mInstance:Ljava/lang/Object; (greylist, reflection, allowed)
    Accessing hidden method Landroid/util/Singleton;->get()Ljava/lang/Object; (greylist, reflection, allowed)
W/cgeo.geocachin: Accessing hidden field Landroid/app/ActivityManager;->IActivityManagerSingleton:Landroid/util/Singleton; (greylist, reflection, allowed)
I/cgeo: [Log] Logging set: minLevel=WARN, minAddCallerInfo=NONE, addCallerInfoMaxDepth=4, throwOnError=false
I/cgeo: [Log] No logging config file 'log-properties.txt' found at /data/user/0/cgeo.geocaching/files/public/BASE/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-file:///data/user/0/cgeo.geocaching/files::/public/BASE/logfiles], using defaults
I/cgeo: [Log] Logging set: minLevel=WARN, minAddCallerInfo=NONE, addCallerInfoMaxDepth=4, throwOnError=false
W/cgeo.geocachin: Accessing hidden field Landroid/view/ViewConfiguration;->sHasPermanentMenuKey:Z (greylist-max-p, reflection, denied)
I/cgeo: [main] [DB] Current Database Version: 97
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 7612
    java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{cgeo.geocaching/cgeo.geocaching.SplashActivity}: java.lang.IllegalAccessException: java.lang.Class<cgeo.geocaching.SplashActivity> is not accessible from java.lang.Class<android.app.AppComponentFactory>
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3194)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalAccessException: java.lang.Class<cgeo.geocaching.SplashActivity> is not accessible from java.lang.Class<android.app.AppComponentFactory>
        at java.lang.Class.newInstance(Native Method)
        at android.app.AppComponentFactory.instantiateActivity(AppComponentFactory.java:95)
        at androidx.core.app.CoreComponentFactory.instantiateActivity(CoreComponentFactory.java:45)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1243)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3182)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
I/Process: Sending signal. PID: 7612 SIG: 9

**Comment by fm-sys** [comment]
Hm, never have seen that exception before. Absolutely no idea. Can anyone else help here?

**Comment by eddiemuc** [comment]
I think class `cgeo.geocaching.SplashActivity` needs to be public. 

See e.g. https://stackoverflow.com/questions/60325094/runtimeexception-java-lang-illegalaccessexception-class-tuseractivity-is-not-a

**Comment by Lineflyer** [comment]
@wolverine007 Are you trying to build your changes or does the problem also occur when building current `master`?

**Comment by wolverine007** [comment]
@Lineflyer Should be the current master, just followed the instructions being stated in the wiki. Only added my Google API key, rest should be still like in current master.

I deleted the Copy, build now a fork and started from scratch... we will see what might happen.

**Comment by bekuno** [comment]
The warning text is ok, no problem. You can compare it with other output on our CI. (https://ci.cgeo.org/)

**Comment by eddiemuc** [comment]
> The app is meanwhile offering to download the routing and map updates in the background, could this process being used to trigger the update of a full local list of caches for instance?

Teammeeting 10.3.24: @wolverine007 since we implemented background cache update in the meantime, is this issue still relevant?

**Comment by wolverine007** [comment]
@eddiemuc
Thanks for the note.
Issue might be solved partly. 

It might be cool, if at least one list could be updated via a scheduled task automatically in the background (e.g. during night hours, when the phone is nornally idle, connected to power and usually to your home wifi - might be checked by the app?).

Or if this is not possible or to complex, a reminder on the home screen might be cool - so a message is shown, when this list has been refreshed x days ago.

Or another idea - highlight the last data refresh inside the cache details view with different colours:

- last refresh <7 days  - white (like it is right now)
- last refresh between 7-14 days - yellow
- last refresh >14 days orange
- last refresh >30 days red

But everything nice to have :)

**Comment by wolverine007** [comment]
@eddiemuc color highlighting from above somehow possible?
BR,
Wolv

**Comment by eddiemuc** [comment]
> [@eddiemuc](https://github.com/eddiemuc) color highlighting from above somehow possible? BR, Wolv

GUI design questions are not exactly my main area of expertise. I would like to redirect this question to maybe @ztNFny or @fm-sys ?
(implementation is possible of course)

**Comment by moving-bits** [comment]
Two ideas: 

For highlighting on cache details page, one could either color the text "stored x days ago" or put a small colored clock symbol beside the text.

If we want to mark entries in cache list, we could add a colored bar at the left, similar to what we have in log list.

If we want to highlight in both places, we should go with the same mechanism, probably colored symbol in this case. For cache list, this could be one of the configurable list items (settings => appearance => configure list info lines)

**Comment by ztNFny** [comment]
Re highlighting: imho there's way more important info, so i don't see why this one needs special attention. In cache list you can easily filter by last refresh date, thanks to stored filters you can easily access that.

Also everything but "now" might be outdated, so the proposed traffic light system doesn't make sense to me. Refreshed 1d ago is not better than refreshed 2 days ago.
